File: 6_tree/力扣加加/路径/非自顶向下/124.二叉树中的最大路径和.ts
import { BinaryTree } from '../../Tree'
import { deserializeNode } from '../../构建类/297二叉树的序列化与反序列化'

/**
 * @param {BinaryTree} root
 * @return {number}
 * @description 自底向上
 * // cur三种路径情况: 1. left+cur 2. right+cur 3. left+cur+right 
   // 其中1,2是要往上探索的。3不能往上。
   // dfs的return值是当前节点[若往上, 即作为子节点]的最大贡献值，是不包含情况3的。
   // 但是3可能是最大路径，因此更新ans时是比较1, 2, 3中最大。
 */
function maxPathSum(root: BinaryTree | null): number {
  if (!root) return 0
  let res = -Infinity

  // 经过root的最大路径长
  const dfs = (root: BinaryTree | null): number => {
    if (!root) return 0
    const leftMax = dfs(root.left)
    const rightMax = dfs(root.right)
    // 和全局变量比较
    res = Math.max(res, root.val + leftMax + rightMax)
    // >0 说明都能使路径变大   类似于求树的高度
    return Math.max(0, Math.max(leftMax, rightMax) + root.val)
  }
  dfs(root)

  return res
}

console.log(maxPathSum(deserializeNode([-10, 9, 20, null, null, 15, 7])!))

File: 7_graph/经典题/路径dfs/124. 二叉树中的最大路径和.py
from collections import defaultdict
from functools import lru_cache
from typing import Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


def treeToGraph(root: TreeNode):
    """二叉树转图 结点id作为键"""

    def dfs(root: Optional[TreeNode], parent: Optional[TreeNode]) -> None:
        if not root:
            return
        rootId = id(root)
        valueById[rootId] = root.val
        if parent is not None:
            parentId = id(parent)
            adjMap[parentId].add(rootId)
            adjMap[rootId].add(parentId)
        dfs(root.left, root)
        dfs(root.right, root)

    adjMap, valueById = defaultdict(set), defaultdict(int)
    dfs(root, None)
    return adjMap, valueById


class Solution:
    def maxPathSum(self, root: TreeNode) -> int:
        @lru_cache(None)
        def dfs(cur: int, parent: int) -> int:
            res = valueById[cur]
            for next in adjMap[cur]:
                if next == parent:
                    continue
                res = max(res, dfs(next, cur) + valueById[cur])
            return res

        adjMap, valueById = treeToGraph(root)
        return max(((dfs(start, -1)) for start in adjMap.keys()), default=root.val)

